### PR TITLE
8270333: -XX:+VerifyStringTableAtExit should not do linear search

### PR DESCRIPTION
**Pull request by iklam**
The implementation of `-XX:+VerifyStringTableAtExit` used a linear search to check for duplicated strings in the interned string table. This would be excessively slow if there were more than a few thousand strings.

This PR uses a hashtable so we can avoid doing a linear search. The table is sized such that each bucket has 8 items on average.

Since this code is executed only for verification, I added a new `java_lang_String::hash_code_noupdate(oop)` function, so we can avoid causing side effects.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270333](https://bugs.openjdk.java.net/browse/JDK-8270333): -XX:+VerifyStringTableAtExit should not do linear search

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 66c4d4ea8502b878168a82c28f918d7d6ea7ffee
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4772/head:pull/4772` \
`$ git checkout pull/4772`

Update a local copy of the PR: \
`$ git checkout pull/4772` \
`$ git pull https://git.openjdk.java.net/jdk pull/4772/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4772`

View PR using the GUI difftool: \
`$ git pr show -t 4772`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4772.diff">https://git.openjdk.java.net/jdk/pull/4772.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/javaClasses.cpp [hunk 1]
```diff
@@ -502,7 +502,8 @@ jchar* java_lang_String::as_unicode_string_or_null(oop java_string, int& length)
   return result;
 }
 
-unsigned int java_lang_String::hash_code(oop java_string) {
+template<bool update>
+unsigned int java_lang_String::hash_code_impl(oop java_string) {
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/javaClasses.cpp:506]: This looks like a really weird way to pass a parameter! Why do we want a template here?

**Review by dholmes-ora — APPROVED** [review]
Changes seem fine, but I don't like your use of templates here.

Thanks,
David

File: src/hotspot/share/classfile/javaClasses.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/classfile/javaClasses.cpp]: Hi David, thanks for the review. I replaced the template function with an inline function.

**Review by dholmes-ora — APPROVED** [review]
Okay ... I'm not sure why you want the code to be duplicated though?

**Comment by iklam** [comment]
> Okay ... I'm not sure why you want the code to be duplicated though?

I could add a new `bool update` parameter to `java_lang_String::hash_code(oop java_string)`, but I worry about performance impacts. 

It probably doesn't matter, but I just don't want to deal with any potential performance issues when changing code used only for debugging.

File: src/hotspot/share/classfile/stringTable.cpp [hunk 2]
```diff
@@ -597,39 +598,40 @@ void StringTable::verify() {
...
+    return java_lang_String::equals(a, b);
+  }
+
+  ResizeableResourceHashtable<oop, bool,
+                              ResourceObj::C_HEAP, mtInternal,
+                              string_hash, string_equals> _table;
  public:
   size_t _errors;
-  VerifyCompStrings(GrowableArray<oop>* oops) : _oops(oops), _errors(0) {}
+  VerifyCompStrings() : _table(unsigned(_items_count + 8 / 8)), _errors(0) {}
```
Comment (yminqi) [inline src/hotspot/share/classfile/stringTable.cpp:613]: I guess you mean unsigned((_items_count + 8) / 8)?

**Review by yminqi — COMMENTED** [review]
Looks good, but one question.

File: src/hotspot/share/classfile/stringTable.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/classfile/stringTable.cpp]: Thanks for the review. I changed it to `_table(unsigned(_items_count / 8) + 1)`

**Review by yminqi — APPROVED** [review]
LGTM.

**Comment by iklam** [comment]
Thanks @dholmes-ora  and @yminqi for the review.
/integrate

**Pull request closed without merging**